### PR TITLE
[Merged by Bors] - feat(Tactic/Linter): the ppRoundtrip linter

### PR DESCRIPTION
**Pull request by adomani**
This linter tries to make sure that the source code looks like the pretty printed version of itself.

It is not intended to be always active, since not every pretty-printed syntax is desirable or type-correct!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+The "pedantic" linter emits a warning when the syntax of a command differs substantially
+from the pretty-printed version of itself.
+
+The linter makes an effort to start the highlighting at the first difference.
+However, it may not always be successful.
+It also prints both the source code and the "expected code" in a 5-character radius from
+the first difference.
+-/
+register_option linter.pedantic : Bool := {
+  defValue := true
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:30]: Didn't you say this should be `false` by default? I assume this is for testing and will be changed before merging.

File: test/simp_confluence.lean [hunk 2]
```diff
@@ -6,6 +6,9 @@ open Function
 
 variable {α β : Type*}
 
+-- it seems that the focusing dot `·` in conv mode pretty-prints without a following space
+set_option linter.pedantic false
+
```
Comment (joneugster) [inline test/simp_confluence.lean:11]: If the linter is turned off by default, this isn't necessary

File: test/PedanticLinter.lean [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import Mathlib.Tactic.Linter.PedanticLinter
+
+set_option linter.pedantic false
+
```
Comment (joneugster) [inline test/PedanticLinter.lean:4]:
```suggestion
```

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  match stx with
+    | .node si k args => .node (capSourceInfo si p) k (args.map (capSyntax · p))
+    | .atom si val => .atom (capSourceInfo si p) (val.take p)
+    | .ident si r v pr => .ident (capSourceInfo si p) { r with stopPos := ⟨min r.stopPos.1 p⟩ } v pr
+    | s => s
+
+namespace Pedantic
+
+/-- Gets the value of the `linter.pedantic` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.pedantic o
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:120]: Same as in another PR: imo it's barely worth when you can just write `Linter.getLinterValue linter.pedantic (← getOptions)` below...

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+For this reason, `polishSource s` performs more conservative changes:
+it only replace all whitespace starting from a linebreak (`\n`) with a single whitespace. -/
+def polishSource (s : String) : String × Array Nat :=
+  let split := s.split (· == '\n')
+  --let lengths := split.map (·.length)
+  let preWS := split.foldl (init := #[]) fun p q =>
+    let txt := q.trimLeft.length
+    (p.push (q.length - txt)).push txt
+  let preWS := preWS.eraseIdx 0
+  --dbg_trace "(spaces, sum, total): {(preWS, preWS.foldl (· + ·) 0, s.length)}"
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:66]: You can probably drop the debug comments now.

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  let s := (split.map .trimLeft).filter (!· == "")
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:67]:
```suggestion
  let s := (split.map .trimLeft).filter (· != "")
```

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+/-- Gets the value of the `linter.pedantic` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.pedantic o
+
+@[inherit_doc Mathlib.Linter.linter.pedantic]
+def pedantic : Linter where run := withSetOptionIn fun stx ↦ do
+    unless getLinterHash (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    let stx:= capSyntax stx (stx.getTailPos?.getD default).1
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:128]:
```suggestion
    let stx := capSyntax stx (stx.getTailPos?.getD default).1
```
can the pedantic linter pedantic-lint itself? ;)

File: Mathlib/Tactic/Linter/PedanticLinter.lean [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+def polishPP (s : String) : String :=
+  let s := s.split (·.isWhitespace)
+  (" ".intercalate (s.filter (!·.isEmpty)))
+    |>.replace "/-!" "/-! "
+    |>.replace "``` " "```  " -- avoid losing an existing space after the triple back-ticks
+                              -- as a consequence of the following replacement
+    |>.replace "`` " "``" -- weird pp ```#eval ``«Nat»``` pretty-prints as ```#eval `` «Nat»```
+    |>.replace "notation3(" "notation3 ("
+    |>.replace "notation3\"" "notation3 \""
+    --|>.replace "{" "{"   -- probably better?
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:54]: ?

**Review by joneugster — APPROVED** [review]
I think this looks good, but as you said it should be disabled unless manually enabled.

File: Mathlib/Tactic/Linter/PedanticLinter.lean (hunk 4 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/PedanticLinter.lean:120]: Update: I think to remember that `getLinterHash` was a problem in other PRs and you should directly call `Linter.getLinterValue linter.pedantic (← getOptions)`, do I remember that correctly?

**Comment by joneugster** [comment]
@adomani: I've pushed my suggestions myself now, adopting some responsibility for this PR, as I think it's a nice tool!

Therefore, somebody else will have to review it. (I might post in the correct Zulip channel about it)

**Comment by joneugster** [comment]
maintainer merge

(please have a bit a deeper look again: the last three commits by myself are essentially unreviewed -- mostly simple fixes)

**Comment by jcommelin** [comment]
Pedantic comment: what do you think of renaming this to something like `ppRoundTrip` instead of `pedantic`? I think that's a bit more descriptive. "pedantic" can be anything, not necessarily formatting-related.

**Comment by jcommelin** [comment]
Otherwise, LGTM

**Comment by adomani** [comment]
Johan, thank you for the comment: I like the new name!

As for the large import, I think that this is to be expected by a linter: it is imported in `Mathlib.Tactic.Linter` and in turn this file is imported virtually everywhere.

**Comment by grunweg** [comment]
Is there anything preventing this linter from landing? All outstanding comments seem to be addressed, CI passes, the linter is disabled by default anyway...

**Comment by adomani** [comment]
I think that all that is missing is a maintainer merging it.  😊

**Comment by grunweg** [comment]
Will this help then? :-)
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**